File: worker/src/api-keys.ts
import { Hono, type MiddlewareHandler } from "hono";
import { requireAuth } from "./auth";
import { consumeRateLimit } from "./rate-limit";
import type { AppEnv, ApiKeyContext, ApiKeyScope } from "./types";

const apiKeys = new Hono<AppEnv>();
const encoder = new TextEncoder();

// Programmatic clients (SDKs, MCP hosts, cron jobs) cannot hold a Firebase ID
// token: those are minted by an interactive sign-in and expire in an hour. An
// API key is the long-lived, per-uid, revocable alternative. Only the SHA-256
// digest is stored, so a database read cannot recover a usable credential.
export const apiKeyPrefix = "omi_sk_";
const keyPattern = /^omi_sk_([0-9a-f]{8})_([A-Za-z0-9_-]{43})$/;
const maximumKeysPerUid = 25;
const lastUsedResolutionMs = 60_000;

export const allScopes: readonly ApiKeyScope[] = [
  "memory:read",
  "currents:read",
  "currents:write",
  "conversations:read",
  "assistant:write",
  "facetime:write",
];

const scopeSet = new Set<string>(allScopes);

const hex = (bytes: Uint8Array) =>
  Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("");

export const digest = async (value: string): Promise<string> =>
  hex(
    new Uint8Array(
      await crypto.subtle.digest("SHA-256", encoder.encode(value)),
    ),
  );

// Compares two same-length hex digests without leaking, through timing, how
// many leading characters matched. Length is not secret (both are SHA-256).
export const timingSafeEqual = (left: string, right: string): boolean => {
  if (left.length !== right.length) return false;
  let difference = 0;
  for (let index = 0; index < left.length; index += 1)
    difference |= left.charCodeAt(index) ^ right.charCodeAt(index);
  return difference === 0;
};

const base64url = (bytes: Uint8Array) => {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary)
    .replaceAll("+", "-")
    .replaceAll("/", "_")
    .replace(/=+$/, "");
};

export const mintApiKey = async (): Promise<{
  key: string;
  prefix: string;
  hash: string;
}> => {
  const prefix = hex(crypto.getRandomValues(new Uint8Array(4)));
  const secret = base64url(crypto.getRandomValues(new Uint8Array(32)));
  const key = `${apiKeyPrefix}${prefix}_${secret}`;
  return { key, prefix, hash: await digest(key) };
};

const parseScopes = (value: unknown): ApiKeyScope[] => {
  if (typeof value !== "string") return [];
  try {
    const parsed = JSON.parse(value) as unknown;
    return Array.isArray(parsed)
      ? (parsed.filter((scope) =>
          scopeSet.has(scope as string),
        ) as ApiKeyScope[])
      : [];
  } catch {
    return [];
  }
};

type KeyRow = {
  id: string;
  uid: string;
  key_hash: string;
  scopes: string;
  email: string | null;
};

export const verifyApiKey = async (
  database: D1Database,
  token: string,
  now: number,
): Promise<{
  uid: string;
  email: string | null;
  key: ApiKeyContext;
} | null> => {
  const parsed = keyPattern.exec(token);
  if (!parsed) return null;
  const candidates = await database
    .prepare(
      `SELECT k.id, k.uid, k.key_hash, k.scopes, u.email
       FROM api_keys k JOIN users u ON u.uid = k.uid
       WHERE k.prefix = ?1 AND k.revoked_at IS NULL
         AND (k.expires_at IS NULL OR k.expires_at > ?2)`,
    )
    .bind(parsed[1], now)
    .all<KeyRow>();
  const presented = await digest(token);
  // Every candidate is compared: no early exit, so a partial prefix collision
  // cannot be distinguished from a miss by response time.
  let matched: KeyRow | null = null;
  for (const row of candidates.results ?? [])
    if (timingSafeEqual(presented, String(row.key_hash))) matched = row;
  if (!matched) return null;
  await database
    .prepare(
      `UPDATE api_keys SET last_used_at = ?1
       WHERE id = ?2 AND (last_used_at IS NULL OR last_used_at < ?3)`,
    )
    .bind(now, matched.id, now - lastUsedResolutionMs)
    .run()
    .catch(() => undefined);
  return {
    uid: String(matched.uid),
    email: matched.email === null ? null : String(matched.email),
    key: { id: String(matched.id), scopes: parseScopes(matched.scopes) },
  };
};

// Accepts either credential on the programmatic surface: an `omi_sk_` API key
// or, unchanged, a Firebase ID token. Firebase-authenticated callers are the
// account owner in person and carry every scope; API keys carry only the
// scopes they were minted with.
export const requireApiAccess: MiddlewareHandler<AppEnv> = async (
  context,
  next,
) => {
  const authorization = context.req.header("authorization") ?? "";
  const bearer = authorization.startsWith("Bearer ")
    ? authorization.slice(7).trim()
    : "";
  const token = bearer || (context.req.header("x-api-key") ?? "").trim();
  if (!token.startsWith(apiKeyPrefix)) return requireAuth(context, next);
  let verified: Awaited<ReturnType<typeof verifyApiKey>>;
  try {
    verified = await verifyApiKey(context.env.DB, token, Date.now());
  } catch {
    return context.json({ error: "Authentication unavailable" }, 503);
  }
  if (!verified) return context.json({ error: "Authentication failed" }, 401);
  context.set("auth", { uid: verified.uid, email: verified.email });
  context.set("apiKey", verified.key);
  await next();
};

export const requireScope =
  (scope: ApiKeyScope): MiddlewareHandler<AppEnv> =>
  async (context, next) => {
    const key = context.get("apiKey");
    if (key && !key.scopes.includes(scope))
      return context.json({ error: "Missing scope", scope }, 403);
    await next();
  };

const object = async (request: Request) => {
  try {
    const value = await request.json();
    return value !== null && typeof value === "object" && !Array.isArray(value)
      ? (value as Record<string, unknown>)
      : null;
  } catch {
    return null;
  }
};

const rowToKey = (row: Record<string, unknown>) => ({
  id: String(row.id),
  name: String(row.name),
  prefix: `${apiKeyPrefix}${String(row.prefix)}`,
  scopes: parseScopes(row.scopes),
  createdAt: Number(row.created_at),
  lastUsedAt: row.last_used_at === null ? null : Number(row.last_used_at),
  expiresAt: row.expires_at === null ? null : Number(row.expires_at),
  revokedAt: row.revoked_at === null ? null : Number(row.revoked_at),
});

apiKeys.get("/", async (context) => {
  const rows = await context.env.DB.prepare(
    `SELECT id, name, prefix, scopes, created_at, last_used_at, expires_at, revoked_at
     FROM api_keys WHERE uid = ?1 ORDER BY created_at DESC LIMIT 100`,
  )
    .bind(context.get("auth").uid)
    .all<Record<string, unknown>>();
  return context.json({ keys: (rows.results ?? []).map(rowToKey) });
});

apiKeys.post("/", async (context) => {
  const body = await object(context.req.raw);
  const name =
    typeof body?.name === "string" &&
    body.name.trim().length > 0 &&
    body.name.length <= 120
      ? body.name.trim()
      : null;
  const requested = body?.scopes === undefined ? [...allScopes] : body.scopes;
  const expiresAt =
    body?.expiresAt === undefined || body.expiresAt === null
      ? null
      : Number(body.expiresAt);
  const now = Date.now();
  if (
    !name ||
    !Array.isArray(requested) ||
    requested.length === 0 ||
    requested.some((scope) => !scopeSet.has(scope as string)) ||
    (expiresAt !== null &&
      (!Number.isSafeInteger(expiresAt) || expiresAt <= now))
  )
    return context.json({ error: "Invalid API key request" }, 400);
  const uid = context.get("auth").uid;
  const limit = await consumeRateLimit(
    context.env,
    `api-key-mint:${uid}`,
    10,
    60 * 60_000,
  );
  if (!limit.allowed)
    return context.json({ error: "Too many requests" }, 429, {
      "retry-after": String(limit.retryAfter),
    });
  const live = await context.env.DB.prepare(
    "SELECT COUNT(*) AS total FROM api_keys WHERE uid = ?1 AND revoked_at IS NULL",
  )
    .bind(uid)
    .first<{ total: number }>();
  if (Number(live?.total ?? 0) >= maximumKeysPerUid)
    return context.json({ error: "API key limit reached" }, 409);
  const scopes = [...new Set(requested as ApiKeyScope[])];
  const minted = await mintApiKey();
  const id = crypto.randomUUID();
  await context.env.DB.prepare(
    `INSERT INTO api_keys (id, uid, name, prefix, key_hash, scopes, created_at, expires_at)
     VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8)`,
  )
    .bind(
      id,
      uid,
      name,
      minted.prefix,
      minted.hash,
      JSON.stringify(scopes),
      now,
      expiresAt,
    )
    .run();
  const stored = await context.env.DB.prepare(
    `SELECT id, name, prefix, scopes, created_at, last_used_at, expires_at, revoked_at
     FROM api_keys WHERE id = ?1 AND uid = ?2`,
  )
    .bind(id, uid)
    .first<Record<string, unknown>>();
  // The plaintext key is returned exactly once; only its digest is retained.
  return context.json({ key: minted.key, apiKey: rowToKey(stored!) }, 201);
});

apiKeys.delete("/:id", async (context) => {
  const revoked = await context.env.DB.prepare(
    "UPDATE api_keys SET revoked_at = ?1 WHERE id = ?2 AND uid = ?3 AND revoked_at IS NULL",
  )
    .bind(Date.now(), context.req.param("id"), context.get("auth").uid)
    .run();
  if (revoked.meta.changes !== 1)
    return context.json({ error: "API key not found" }, 404);
  return context.body(null, 204);
});

export default apiKeys;

File: worker-rs/src/api_keys.rs
//! Pure parity port of `worker/src/api-keys.ts`.
//!
//! Programmatic clients cannot hold a Firebase ID token, so an `omi_sk_` API
//! key is the long-lived, per-uid, revocable alternative. Only the SHA-256
//! digest is ever stored; a database read cannot recover a usable credential.
//!
//! Everything here is decision logic and is host-testable; the D1 lookup and
//! the HTTP routes live in `routes_keys` (wasm only).

use serde_json::Value;

use crate::crypto_util::{base64url, constant_time_eq, sha256_hex, to_hex_lower};
use crate::jsnum::{is_safe_integer, number_from_value};

pub const API_KEY_PREFIX: &str = "omi_sk_";
pub const MAXIMUM_KEYS_PER_UID: i64 = 25;
pub const LAST_USED_RESOLUTION_MS: i64 = 60_000;
/// `consumeRateLimit(env, "api-key-mint:<uid>", 10, 60 * 60_000)`.
pub const MINT_RATE_LIMIT: i64 = 10;
pub const MINT_RATE_WINDOW_MS: i64 = 60 * 60_000;
pub const NAME_MAX_CHARACTERS: usize = 120;

/// `allScopes` — the closed set of scopes a key may carry.
pub const ALL_SCOPES: &[&str] = &[
    "memory:read",
    "currents:read",
    "currents:write",
    "conversations:read",
    "assistant:write",
    "facetime:write",
    "speech:write",
];

/// `scopeSet.has(scope)`.
pub fn is_scope(value: &str) -> bool {
    ALL_SCOPES.contains(&value)
}

/// SHA-256 → lowercase hex, matching the TS `digest`.
pub fn digest(value: &str) -> String {
    sha256_hex(value)
}

/// Constant-time hex-digest comparison (`timingSafeEqual`).
pub fn timing_safe_equal(left: &str, right: &str) -> bool {
    constant_time_eq(left, right)
}

/// `keyPattern = /^omi_sk_([0-9a-f]{8})_([A-Za-z0-9_-]{43})$/` — returns the
/// public prefix capture when the token is well formed.
pub fn parse_key(token: &str) -> Option<&str> {
    let rest = token.strip_prefix(API_KEY_PREFIX)?;
    let (prefix, secret) = rest.split_once('_')?;
    if prefix.len() != 8
        || !prefix
            .bytes()
            .all(|b| b.is_ascii_hexdigit() && !b.is_ascii_uppercase())
    {
        return None;
    }
    if secret.len() != 43
        || !secret
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b == b'_' || b == b'-')
    {
        return None;
    }
    Some(prefix)
}

/// `parseScopes` — a JSON-encoded array column, filtered to the known set.
/// Anything unparsable or non-array yields an empty list.
pub fn parse_scopes(value: Option<&Value>) -> Vec<String> {
    let Some(Value::String(raw)) = value else {
        return Vec::new();
    };
    let Ok(Value::Array(items)) = serde_json::from_str::<Value>(raw) else {
        return Vec::new();
    };
    items
        .iter()
        .filter_map(Value::as_str)
        .filter(|scope| is_scope(scope))
        .map(str::to_string)
        .collect()
}

/// A candidate row from the prefix lookup.
pub struct KeyCandidate {
    pub id: String,
    pub uid: String,
    pub key_hash: String,
    pub scopes: Vec<String>,
    pub email: Option<String>,
}

/// Selects the row whose stored digest matches the presented one.
///
/// Every candidate is compared — there is no early exit — so a partial prefix
/// collision cannot be distinguished from a miss by response time. The last
/// match wins, matching the TS loop exactly.
pub fn select_match<'a>(
    presented: &str,
    candidates: &'a [KeyCandidate],
) -> Option<&'a KeyCandidate> {
    let mut matched: Option<&KeyCandidate> = None;
    for row in candidates {
        if timing_safe_equal(presented, &row.key_hash) {
            matched = Some(row);
        }
    }
    matched
}

/// The credential presented on the programmatic surface.
pub enum Credential {
    /// An `omi_sk_` API key: verify against D1.
    ApiKey(String),
    /// Anything else (including nothing): fall through to Firebase `requireAuth`.
    Firebase,
}

/// `requireApiAccess`'s credential selection: `Authorization: Bearer …` first,
/// then `x-api-key`. Only a token starting with the `omi_sk_` prefix takes the
/// API-key path.
pub fn credential(authorization: &str, x_api_key: &str) -> Credential {
    let bearer = authorization
        .strip_prefix("Bearer ")
        .map(str::trim)
        .unwrap_or("");
    let token = if bearer.is_empty() {
        x_api_key.trim()
    } else {
        bearer
    };
    if token.starts_with(API_KEY_PREFIX) {
        Credential::ApiKey(token.to_string())
    } else {
        Credential::Firebase
    }
}

/// A freshly minted key: the plaintext (returned exactly once), the public
/// prefix, and the digest that is all the database ever holds.
pub struct MintedKey {
    pub key: String,
    pub prefix: String,
    pub hash: String,
}

/// `mintApiKey` with the randomness injected so it is testable.
pub fn mint_api_key(prefix_bytes: [u8; 4], secret_bytes: [u8; 32]) -> MintedKey {
    let prefix = to_hex_lower(&prefix_bytes);
    let secret = base64url(&secret_bytes);
    let key = format!("{API_KEY_PREFIX}{prefix}_{secret}");
    let hash = digest(&key);
    MintedKey { key, prefix, hash }
}

/// A validated `POST /v1/api-keys` body.
pub struct MintRequest {
    pub name: String,
    pub scopes: Vec<String>,
    pub expires_at: Option<i64>,
}

/// Validation for `POST /v1/api-keys`. `None` is the TS
/// `{ error: "Invalid API key request" }, 400`.
pub fn validate_mint(body: Option<&Value>, now: i64) -> Option<MintRequest> {
    let name = match body.and_then(|b| b.get("name")) {
        Some(Value::String(raw)) => {
            let trimmed = raw.trim();
            (!trimmed.is_empty() && raw.chars().count() <= NAME_MAX_CHARACTERS)
                .then(|| trimmed.to_string())
        }
        _ => None,
    }?;

    let requested: Vec<String> = match body.and_then(|b| b.get("scopes")) {
        None => ALL_SCOPES.iter().map(|s| s.to_string()).collect(),
        Some(Value::Array(items)) => {
            if items.is_empty() || items.iter().any(|i| !i.as_str().is_some_and(is_scope)) {
                return None;
            }
            items
                .iter()
                .filter_map(Value::as_str)
                .map(str::to_string)
                .collect()
        }
        // `scopes: null` (or any non-array) is not an array: rejected.
        Some(_) => return None,
    };

    let expires_at = match body.and_then(|b| b.get("expiresAt")) {
        None | Some(Value::Null) => None,
        Some(value) => {
            let parsed = number_from_value(value);
            if !is_safe_integer(parsed) || parsed <= now as f64 {
                return None;
            }
            Some(parsed as i64)
        }
    };

    // `[...new Set(requested)]` — de-duplicated, insertion order preserved.
    let mut scopes: Vec<String> = Vec::new();
    for scope in requested {
        if !scopes.contains(&scope) {
            scopes.push(scope);
        }
    }
    Some(MintRequest {
        name,
        scopes,
        expires_at,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn candidate(id: &str, uid: &str, hash: &str) -> KeyCandidate {
        KeyCandidate {
            id: id.into(),
            uid: uid.into(),
            key_hash: hash.into(),
            scopes: vec!["memory:read".into()],
            email: None,
        }
    }

    #[test]
    fn mint_produces_a_pattern_matching_key_and_stores_only_the_digest() {
        let minted = mint_api_key([0xde, 0xad, 0xbe, 0xef], [7u8; 32]);
        assert_eq!(minted.prefix, "deadbeef");
        assert!(minted.key.starts_with("omi_sk_deadbeef_"));
        assert_eq!(parse_key(&minted.key), Some("deadbeef"));
        assert_eq!(minted.hash, digest(&minted.key));
        assert!(!minted.hash.contains(&minted.key));
    }

    #[test]
    fn rejects_malformed_keys() {
        for token in [
            "",
            "omi_sk_",
            "omi_sk_deadbeef",
            "omi_sk_DEADBEEF_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            "omi_sk_deadbee_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            // 42 secret characters, one short.
            "omi_sk_deadbeef_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            "omi_sk_deadbeef_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa+",
            "sk_deadbeef_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        ] {
            assert!(parse_key(token).is_none(), "should reject {token}");
        }
    }

    #[test]
    fn compares_digests_without_early_exit() {
        let real = digest("omi_sk_deadbeef_real");
        let forged = digest("omi_sk_deadbeef_forged");
        let rows = [
            candidate("a", "uid-a", &forged),
            candidate("b", "uid-b", &real),
        ];
        assert_eq!(select_match(&real, &rows).map(|r| r.id.as_str()), Some("b"));
        // A prefix collision with a forged secret matches nothing.
        assert!(select_match(&digest("omi_sk_deadbeef_other"), &rows).is_none());
        assert!(select_match(&real, &[]).is_none());
    }

    #[test]
    fn forged_secret_with_a_real_prefix_is_rejected() {
        let minted = mint_api_key([1, 2, 3, 4], [9u8; 32]);
        let rows = [candidate("k", "uid", &minted.hash)];
        let forged = format!("{}x", &minted.key[..minted.key.len() - 1]);
        assert_eq!(parse_key(&forged), parse_key(&minted.key));
        assert!(select_match(&digest(&forged), &rows).is_none());
        assert!(select_match(&digest(&minted.key), &rows).is_some());
    }

    #[test]
    fn parse_scopes_drops_unknown_and_malformed_values() {
        assert_eq!(
            parse_scopes(Some(&json!(r#"["memory:read","nope","currents:write"]"#))),
            vec!["memory:read".to_string(), "currents:write".to_string()]
        );
        assert!(parse_scopes(Some(&json!("not json"))).is_empty());
        assert!(parse_scopes(Some(&json!(r#"{"a":1}"#))).is_empty());
        assert!(parse_scopes(Some(&Value::Null)).is_empty());
        assert!(parse_scopes(None).is_empty());
    }

    #[test]
    fn credential_prefers_bearer_then_x_api_key() {
        assert!(matches!(
            credential("Bearer omi_sk_abc", ""),
            Credential::ApiKey(token) if token == "omi_sk_abc"
        ));
        assert!(matches!(
            credential("", "  omi_sk_abc  "),
            Credential::ApiKey(token) if token == "omi_sk_abc"
        ));
        assert!(matches!(
            credential("Bearer eyJhbGciOi", ""),
            Credential::Firebase
        ));
        assert!(matches!(credential("", ""), Credential::Firebase));
    }

    #[test]
    fn mint_validation_defaults_to_every_scope() {
        let request = validate_mint(Some(&json!({ "name": "  laptop  " })), 1_000).unwrap();
        assert_eq!(request.name, "laptop");
        assert_eq!(request.scopes.len(), ALL_SCOPES.len());
        assert!(request.expires_at.is_none());
    }

    #[test]
    fn mint_validation_rejects_unknown_scopes_and_empty_lists() {
        for body in [
            json!({ "name": "k", "scopes": [] }),
            json!({ "name": "k", "scopes": ["memory:read", "admin"] }),
            json!({ "name": "k", "scopes": "memory:read" }),
            json!({ "name": "k", "scopes": null }),
            json!({ "name": "   " }),
            json!({ "scopes": ["memory:read"] }),
            json!({ "name": 12 }),
        ] {
            assert!(
                validate_mint(Some(&body), 1_000).is_none(),
                "should reject {body}"
            );
        }
        assert!(validate_mint(None, 1_000).is_none());
    }

    #[test]
    fn mint_validation_rejects_a_past_or_unsafe_expiry() {
        assert!(validate_mint(Some(&json!({ "name": "k", "expiresAt": 999 })), 1_000).is_none());
        assert!(validate_mint(Some(&json!({ "name": "k", "expiresAt": 1_000 })), 1_000).is_none());
        assert!(validate_mint(Some(&json!({ "name": "k", "expiresAt": 1.5 })), 1_000).is_none());
        assert!(validate_mint(Some(&json!({ "name": "k", "expiresAt": "nope" })), 1_000).is_none());
        assert_eq!(
            validate_mint(Some(&json!({ "name": "k", "expiresAt": 2_000 })), 1_000)
                .unwrap()
                .expires_at,
            Some(2_000)
        );
        assert!(
            validate_mint(Some(&json!({ "name": "k", "expiresAt": null })), 1_000)
                .unwrap()
                .expires_at
                .is_none()
        );
    }

    #[test]
    fn mint_validation_deduplicates_scopes() {
        let request = validate_mint(
            Some(&json!({ "name": "k", "scopes": ["memory:read", "memory:read"] })),
            0,
        )
        .unwrap();
        assert_eq!(request.scopes, vec!["memory:read".to_string()]);
    }

    #[test]
    fn candidate_email_is_carried_through() {
        let row = KeyCandidate {
            email: Some("a@b.co".into()),
            ..candidate("i", "u", "h")
        };
        assert_eq!(row.email.as_deref(), Some("a@b.co"));
        assert_eq!(row.scopes, vec!["memory:read".to_string()]);
    }
}
